src/python/pants/engine/internals/engine_test.py

```diff
@@ -392,10 +393,41 @@ def test_streaming_workunits_parent_id_and_rule_metadata(self):
         r3 = next(item for item in finished if item["name"] == "rule_three")
         r4 = next(item for item in finished if item["name"] == "rule_four")
 
-        assert r1.get("parent_id", None) is None
+        # rule_one should have a parent ID that is not emitted because of its low log level
```

Comment:
So, thanks for writing this test (and including this comment!) because this is an interesting case that we're going to have to deal with both here and anywhere consuming these:

Whenever workunits are "filtered out" like this, we almost certainly want to fix up the parent link so that it is not broken. By that I mean: we should walk the workunit graph until we encounter a parent that _isn't_ filtered out, and make that the new parent of the workunit.

From a "this is a tree" perspective, filtering intermediate nodes still results in a tree: and that's important for consumers of this data.